lib/src/SessionManager.cc

```diff
@@ -16,50 +16,92 @@
 
 using namespace drogon;
 
-SessionManager::SessionManager(trantor::EventLoop *loop, size_t timeout)
-    : loop_(loop), timeout_(timeout)
+class CacheMapSessionStorageProvider : public SessionStorageProvider
 {
-    assert(timeout_ >= 0);
-    if (timeout_ > 0)
+  public:
+    CacheMapSessionStorageProvider(trantor::EventLoop *loop,
+                                   float tickInterval,
+                                   size_t wheelsNum,
+                                   size_t bucketsNumPerWheel)
+        : data_(std::make_unique<CacheMap<std::string, SessionPtr>>(
+              loop,
+              tickInterval,
+              wheelsNum,
+              bucketsNumPerWheel))
+    {
+    }
+
+    bool findAndFetch(const std::string &key, SessionPtr &value) override
+    {
+        return data_->findAndFetch(key, value);
+    }
+
+    void insert(const std::string &key,
+                const SessionPtr &value,
+                size_t timeout) override
+    {
+        data_->insert(key, value, timeout);
+    }
+
+  private:
+    std::unique_ptr<CacheMap<std::string, SessionPtr>> data_;
+};
+
+SessionManager::SessionManager(std::unique_ptr<SessionStorageProvider> provider,
+                               size_t timeout)
+    : provider_(std::move(provider)), timeout_(timeout)
+{
+}
+
+SessionPtr SessionManager::getSession(const std::string &sessionID,
+                                      bool needToSet)
+{
+    assert(!sessionID.empty());
+    SessionPtr sessionPtr;
+    std::lock_guard<std::mutex> lock(mapMutex_);
+    if (!provider_->findAndFetch(sessionID, sessionPtr))
+    {
+        sessionPtr = std::make_shared<Session>(sessionID, needToSet);
+        provider_->insert(sessionID, sessionPtr, timeout_);
```

Comment:
If the provider_ is created for redis, how do we store the sessionPtr to the redis server?
And as the session object (sessionPtr) can be accessed by user via Http Request objects, how do we synchronize its changes to the redis server?

Reply:
I think we have to make the Session class as a pure virtual class for different implementations of memory storage or redis storage.

Reply:
Protobuf for session fields serialization/deserialization?